In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

In [2]:
csv_files=['social_media.csv','security_logs.csv','relationships.csv','person.csv','location.csv','location_history.csv','interview.csv',
          'forensic_evidence.csv','financial_transactions','employment.csv','drivers_license.csv','digital_evidence.csv','crime_scene_report.csv',
          'company.csv']

In [3]:
#pd.set_option('display.max_colwidth', None)  # Don't truncate strings
#pd.set_option('display.max_columns', None)   # Show all columns
#pd.set_option('display.width', 1000)         # Wide enough so pandas doesn't wrap lines


<h2 style="color :red">Murder was done on 15th june at 8:30PM</h2>

In [4]:
df = pd.read_csv('crime_scene_report.csv') 
# Convert the 'date' column to datetime
df['date'] = pd.to_datetime(df['date'], format='%Y%m%d')
filtered = df[
    (df['type'].str.lower() == 'murder') &
    (df['date'] == pd.to_datetime('2023-06-15'))
]
pd.set_option('display.max_colwidth', None)


print(filtered['description'])

0    Tech billionaire Elijah Armstrong found dead in his penthouse office. Multiple toxins detected, suggesting a sophisticated assassination. Security systems showed evidence of quantum-level tampering. Victim was involved in controversial merger with implications for international quantum computing research. Last known video shows meeting with unknown diplomatic personnel. Project Phoenix files partially deleted.
Name: description, dtype: object


<h3 style="color:green"></br>
💀 Victim: Elijah Armstrong</br>
Date of death: June 15, 2023</br>
Cause: Multiple toxins (tetrodotoxin, synthetic digitalis, rare polonium isotope)</br>
Time window: 20:15–20:45</br>
Crime scene: His penthouse office</br>
Key clue: Involvement in Project Phoenix and an international quantum computing merger</h3>

<h3 style="color:red">DIGITAL_EVIDENCE</h3>

In [5]:
digital_df=pd.read_csv('digital_evidence.csv')
digital_df['formatted_timestamp'] = pd.to_datetime(digital_df['timestamp'], unit='s')
result=digital_df[['person_id','device_type','content_type','formatted_timestamp','content']]
#result = result.sort_values(by='formatted_timestamp')
pd.set_option('display.max_colwidth', None)
print(result)

    person_id          device_type       content_type formatted_timestamp  \
0          10               Laptop              Email 2023-06-15 16:02:00   
1          10         Mobile Phone       Text Message 2023-06-15 22:10:00   
2          10      Office Computer    Browser History 2023-06-15 21:30:00   
3          10      Building System         Access Log 2023-06-15 23:13:00   
4          30               Tablet           Document 2023-06-15 20:02:00   
5          40         Mobile Phone       Text Message 2023-06-15 22:10:00   
6          50      Security System           Log File 2023-06-15 22:35:00   
7          80               Laptop  Remote Access Log 2023-06-15 22:10:00   
8          10               Server   Financial Record 2023-06-14 16:00:00   
9          20         Mobile Phone           Calendar 2023-06-15 16:02:00   
10        110     Quantum Computer           Log File 2023-06-15 22:35:00   
11        110       Quantum Server       Research Log 2023-10-03 18:30:00   

<h3>Digital Evidence Summary</h3>

<h5 style="color:blue">On June 15, 2023, Person 10 received the final merger documents from Maya, which included mention of a contingency plan, and later that evening postponed a meeting citing integration issues with Project Phoenix. Their browser activity around 9:30 PM revealed searches related to the merger, contingency planning, and rare toxin markets, while security logs show they unlocked their office at 7:55 PM, locked it from inside a minute later, and multiple failed emergency override attempts were made around 8:35 PM—only succeeding at 9:53 PM using the Security Director's credentials, after the secondary security system had been disabled at 8:05 PM using admin credentials (vtk). Earlier that evening, at 8:02 PM, Person 30 filed a confidential report exposing a backdoor vulnerability in Project Phoenix. Later, at 10:35 PM, Person 50 triggered camera maintenance, during which encrypted data was transferred to NexusTech. Months later, on October 3, 2023, Person 110 reported a critical quantum vulnerability requiring diplomatic intervention, followed by Person 160 sending an encrypted diplomatic message that included authorization for elimination.</h5>

<h2 style="color:Red">Location Hist</h2>

In [6]:
location_hist_df=pd.read_csv('location_history.csv')
location_hist_df['formatted_timestamp'] = pd.to_datetime(digital_df['timestamp'], unit='s')
def convert_seconds(seconds):
    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    secs = seconds % 60
    return f"{int(hours)}h {int(minutes)}m {int(secs)}s"


location_hist_df['duration_formatted'] = location_hist_df['duration'].apply(convert_seconds)
result=location_hist_df[['person_id','location_id','formatted_timestamp','duration_formatted','activity']]
result = result.sort_values(by='formatted_timestamp')
print(result)


    person_id  location_id formatted_timestamp duration_formatted  \
8          80           13 2023-06-14 16:00:00          10h 0m 0s   
0          10            1 2023-06-15 16:02:00          3h 30m 0s   
9          90            1 2023-06-15 16:02:00           4h 0m 0s   
4          40           12 2023-06-15 20:02:00           2h 0m 0s   
2          30            3 2023-06-15 21:30:00           3h 0m 0s   
1          20            5 2023-06-15 22:10:00           5h 0m 0s   
5          40            1 2023-06-15 22:10:00           2h 0m 0s   
7          60            1 2023-06-15 22:10:00          5h 30m 0s   
6          50            1 2023-06-15 22:35:00           6h 0m 0s   
10         90            7 2023-06-15 22:35:00           3h 0m 0s   
3          30           11 2023-06-15 23:13:00          11h 0m 0s   
11        100            8 2023-10-03 18:30:00          2h 30m 0s   
12         60            8 2023-10-03 18:35:00          0h 30m 0s   
13        100            5 2023-10

In [7]:
location_df = pd.read_csv('location.csv')
filtered = location_df[
    (location_df['id'] == 1) |
    (location_df['id'] == 12) |
    (location_df['id'] == 3) |
    (location_df['id'] == 5) |
    (location_df['id'] == 11)
]
print(filtered)

    id                    name                address      city  \
0    1         CryptoSphere HQ    1000 Innovation Way  SQL City   
2    3    CryptoSphere R&D Lab    1050 Innovation Way  SQL City   
4    5         The Grand Hotel           888 Broadway  SQL City   
10  11  Viktor Radek Residence  1200 Marina Boulevard  SQL City   
11  12    Sarah Chen Residence       42 Sunset Avenue  SQL City   

                 type  latitude  longitude  
0     Office Building   37.7749  -122.4194  
2   Research Facility   37.7749  -122.4196  
4               Hotel   37.7879  -122.4074  
10          Residence   37.8035  -122.4399  
11          Residence   37.7459  -122.4863  


<h4 style="color:red">Summary Of Location History</h4>
<h5 style="color:blue">Person 30 has a highly inconsistent alibi—claiming to be "Working in lab" until 00:30 and then suddenly "At residence" (Viktor Radek's home)</br> at 23:13 for 11 hours, which is impossible—suggesting a fabricated timeline. Person 90 and Person 100 also have conflicting location records, while</br> The Grand Hotel (location_id 5) appears in multiple suspicious activities. The most likely scenario is that Person 30 lied about their </br>whereabouts, possibly to cover up involvement in a crime (such as a murder) at Viktor Radek's residence or another location during the time </br>discrepancy. Security patrol overlaps and other suspicious movements (like Person 100's impossible work/executive club timeline) may indicate coordinated deception or witness tampering.</h5>

<h2 style="color:red">CCTV photage of 15th june</h2>

In [8]:
security_df=pd.read_csv('security_logs.csv')
security_df['date'] = pd.to_datetime(security_df['date'], format='%Y%m%d')
filtered = security_df[
    (security_df['date'] == pd.to_datetime('2023-06-15'))
]
result=filtered[['person_id','date','time','action','details']]
pd.set_option('display.max_colwidth', None)
print(result)

    person_id       date  time          action  \
0        10.0 2023-06-15  1705           Enter   
1        60.0 2023-06-15  1832            Exit   
2        30.0 2023-06-15  1845           Enter   
3        90.0 2023-06-15  1955            Exit   
4        30.0 2023-06-15  2012            Exit   
5        50.0 2023-06-15  2015           Enter   
6        50.0 2023-06-15  2025            Exit   
7        50.0 2023-06-15  2026           Enter   
8        50.0 2023-06-15  2048            Exit   
9        50.0 2023-06-15  2049           Enter   
10       50.0 2023-06-15  2110            Exit   
11       50.0 2023-06-15  2111           Enter   
12       50.0 2023-06-15  2135            Exit   
13       50.0 2023-06-15  2136           Enter   
14       50.0 2023-06-15  2158            Exit   
15       40.0 2023-06-15  2147           Enter   
16       40.0 2023-06-15  2153  Access Attempt   
17       50.0 2023-06-15  2153        Override   
18       50.0 2023-06-15  2155  Access Granted   


<h2>Summary of CCTV PHOTAGE</h2>

<h5 style="color:blue">On the evening in question, Person 10 arrived at 5:05 PM via normal security checks with digital activity confirming their presence, Person 30 entered at 6:45 PM and left at 8:12 PM after using an admin override to access Armstrong’s office—possibly to report or assist with an issue, Person 50 showed highly suspicious behavior by frequently moving between the 45th and 50th floors from 8:15 PM to 9:58 PM, including using an emergency override to enter the 50th-floor office amidst strange digital activity suggesting tampering, Person 40 arrived at 9:47 PM but was initially blocked from the 50th floor due to a lockdown until Person 50 granted them access, while Persons 60 and 90 entered and exited without any unusual activity.</h5>

<h2 style="color:red">Forensic Report</h2>

In [9]:
forensic_df = pd.read_csv('forensic_evidence.csv')


filtered = forensic_df[forensic_df['crime_scene_report_id'] == 1]

# Select only the person_id and description columns
result = filtered[['person_id','type', 'description']]

# Show the result
print(result)

   person_id          type  \
0       10.0    Toxicology   
1       30.0  Fingerprints   
2        NaN    Electronic   
3       10.0      Document   
4       50.0         Trace   
5        NaN       Digital   
6        NaN    Biological   
7       30.0       Residue   
8       30.0       Digital   

                                                                                                                                                                                                                                                                                                description  
0                                 Multiple toxins detected: traces of tetrodotoxin in victim's coffee cup (20:15 sample), synthetic digitalis compound in water glass (19:45 sample), and rare polonium isotope residue on desk surface (source: NexusTech lab). Death window revised: between 20:15-20:45.  
1                                                                 Multiple fingerprints found: 

<h3>Summary of forensic Report</h3>

<h5 style="color:blue">The victim was poisoned with rare toxins—tetrodotoxin in coffee, digitalis in water, and polonium on the desk—suggesting a diplomatic or high-level plot; fingerprints, DNA, and clothing fibers link multiple suspects including Viktor Radek, S. Zhang, and a mysterious diplomat, while tampered security, stolen Project Phoenix files, and traces of diplomatic-grade materials point to a coordinated attack involving diplomatic access and possibly Chinese and Russian embassy ties.</h5>

In [10]:

rel_df = pd.read_csv('relationships.csv')

related = rel_df[(rel_df['person1_id'] == 10) | (rel_df['person2_id'] == 10)
|(rel_df['person1_id'] == 30) | (rel_df['person2_id'] == 30)
|(rel_df['person1_id'] == 50) | (rel_df['person2_id'] == 50)
|(rel_df['person1_id'] == 60) | (rel_df['person2_id'] == 60)
|(rel_df['person1_id'] == 90) | (rel_df['person2_id'] == 90)
]

print(related)


    id  person1_id  person2_id relationship_type  start_date  end_date  \
0    1          10          20      Professional    20160712       NaN   
1    2          10          30      Professional    20170305       NaN   
2    3          10          40      Professional    20180110       NaN   
3    4          10          60      Professional    20210908       NaN   
4    5          10          70      Professional    20150403       NaN   
5    6          10         100      Professional    20220105       NaN   
6    7          20          30          Personal    20180610       NaN   
7    8          30          50      Professional    20190614       NaN   
8    9          30          80      Professional    20230511       NaN   
9   10          40          10          Personal    20220304       NaN   
10  11          50          80      Professional    20230511       NaN   
11  12          90          30      Professional    20211105       NaN   
13  14         110          30      Pr

<h2 style="color:red">SUSPECTS SEEMS TO INVOLVED IN MURDER</h2>

In [11]:
person_df=pd.read_csv('person.csv')
filtered = person_df[(person_df['id'] == 10) | (person_df['id'] == 30) | (person_df['id'] == 50)|(person_df['id'] == 40)|
(person_df['id'] == 60) |(person_df['id'] == 90)]
result=filtered['name']
print(result)

0    Elijah Armstrong
2        Viktor Radek
3          Sarah Chen
4       Marcus Taylor
5      Emma Rodriguez
8        David Okafor
Name: name, dtype: object


<h2 style="color:red">Financial_Transaction</h2>

In [12]:
financial_df=pd.read_csv('financial_transactions.csv')
financial_df.head()

id  person_id     transaction_type      amount      date  time  \
0   1         30        Wire Transfer  2500000.00  20230610  1425   
1   2         30      Cash Withdrawal    25000.00  20230612  1055   
2   3         20      Restaurant Bill     1247.86  20230615  2243   
3   4         40       Stock Purchase   350000.00  20230614  1010   
4   5         10  Real Estate Payment  4500000.00  20230603   930   

   location_id  recipient_id account_number  \
0          NaN           NaN  CAYM182736495   
1          8.0           NaN   CRYP78291046   
2          5.0           NaN   CRYP24680135   
3          NaN           NaN   CRYP13572468   
4          NaN           NaN   CRYP97531086   

                             transaction_details  
0                   Transfer to offshore account  
1                       Maximum daily withdrawal  
2   Dinner with 6 guests - QuantumLink investors  
3            Purchase of RivalTech stock options  
4  Final payment on diplomatic district property

In [13]:
print(financial_df)

    id  person_id       transaction_type      amount      date  time  \
0    1         30          Wire Transfer  2500000.00  20230610  1425   
1    2         30        Cash Withdrawal    25000.00  20230612  1055   
2    3         20        Restaurant Bill     1247.86  20230615  2243   
3    4         40         Stock Purchase   350000.00  20230614  1010   
4    5         10    Real Estate Payment  4500000.00  20230603   930   
5    6         50        Cash Withdrawal     5000.00  20230613  1622   
6    7         50               Purchase     3499.99  20230614  1245   
7    8         30          Wire Transfer  1000000.00  20230614  1630   
8    9         70  Expense Reimbursement    12456.78  20230614   945   
9   10        100        Restaurant Bill      342.56  20230615  2035   
10  11        110          Wire Transfer   750000.00  20230612  1515   
11  12         20         Cryptocurrency  1500000.00  20230615  1922   
12  13         80                Payment    25000.00  20230614  

<h4 style="color:red">Summary of financial report</h4>
<h5>Viktor Radek (primary executor) – Moved millions offshore before the murder.</br>
</br>
Sarah Chen (financial accomplice) – Betrayed Armstrong by investing in rivals.</br>
</br>
Marcus Taylor (security insider) – Purchased specialized equipment used in the crime.</br>
</br>
Diplomatic Entities – Funded the operation and covered it up post-murder.</h5>


In [16]:
interview_df=pd.read_csv('interview.csv')
print(interview_df)

    id  person_id  \
0    1         60   
1    2         40   
2    3         30   
3    4         50   
4    5         20   
5    6         70   
6    7         80   
7    8         90   
8    9        100   
9   10        110   
10  11        130   
11  12        140   
12  13        150   
13  14        160   
14  15        170   
15  16        180   
16  17        190   
17  18         60   

                                                                                                                                                                                                                                                                                                                                                 transcript  \
0      Mr. Armstrong asked me to reschedule his dinner plans on the evening of June 15th. He mentioned needing to review merger documents with Sarah before tomorrow's announcement. He seemed tense but not unusually so. I left his office around 6:30 

<h4 style="color:blue">Viktor is prime suspect #1—foreign ties, shady chat, BMW getaway
Security Guard (#3) or New Hire (#6)—they had access and opportunity.
And Quantum Researcher (#9)Either a victim… or a mastermind.</h4>

In [17]:
driver_df=pd.read_csv('drivers_license.csv')
driver_df.head()

id  age  height eye_color hair_color  gender plate_number  car_make  \
0  10001   52     188      blue       gray    male       CS5290     Tesla   
1  10002   45     175     hazel      black  female       NXS718       BMW   
2  10003   38     182     brown      brown    male       QNT546      Audi   
3  10004   41     168     green        red  female       EVC999  Mercedes   
4  10005   47     180     brown      black    male       BLK357     Dodge   

  car_model  car_year  
0   Model X      2022  
1        X5      2020  
2        A8      2021  
3   S-Class      2022  
4   Charger      2019

In [18]:
print(driver_df)

       id  age  height eye_color     hair_color  gender plate_number  \
0   10001   52     188      blue           gray    male       CS5290   
1   10002   45     175     hazel          black  female       NXS718   
2   10003   38     182     brown          brown    male       QNT546   
3   10004   41     168     green            red  female       EVC999   
4   10005   47     180     brown          black    male       BLK357   
5   10006   35     170     brown         blonde  female       CRY123   
6   10007   52     175      blue  salt & pepper    male       TEC007   
7   10008   29     165      blue         blonde  female       SEC808   
8   10009   32     177     brown          brown    male       DEV404   
9   10010   46     183     green          brown    male       FIN555   
10  10011   38     165     brown          black  female       QNT123   
11  10012   45     182      blue          brown    male       RES456   
12  10013   35     175     hazel         blonde  female       SC

print(driver_df)

<h1 style="color:red">!FINAL CONCLUSION</h1>

<h4 style="color:blue">KEY SUSPECT AND THEIR ROLES</h4>
<h5>Viktor Radek (Person ID: 30)</br>
Motive: Financial and professional grievances. Forensic reports show he was involved in a controversial merger that reduced his ownership stake. Financial transactions reveal a $2.5 million offshore transfer days before the murder.</br>
</br>
Opportunity: He was in the R&D lab until 9 PM but used an executive override to access Armstrongs office at 8:03 PM. His car (BMW X5) matches witness accounts of a getaway vehicle.</br>
</br>
Evidence: His fingerprints were found on Armstrong's laptop, and his suit fibers matched toxin residue.</br> A witness overheard him speaking Russian about a "package" and "Armstrong" the day before the murder.</br>
</h5>

<h4>Sarah Chen (Person ID: 40)</h4>
<h5>Motive: Close ties to Armstrong but had financial conflicts. She purchased RivalTech stock options before the murder, suggesting insider trading.</br>
</br>
Opportunity: She was supposed to meet Armstrong at 8:30 PM but was delayed. Security logs show her entering the building at 9:47 PM, granted access by</br> a security override.</br>
</br>
Evidence: Her fingerprints were found on a document cabinet, and she received a suspicious text: "Remember our deal. The merger must not proceed."</h5>

<h4>Marcus Taylor (Security Guard, Person ID: 50)</h4>
<h5>Motive: Possible coercion or bribery. His financial records show a $15,000 payment labeled "special security equipment - diplomatic grade."</br>
</br>
Opportunity: He was patrolling floors 45-50 during the murder window and disabled cameras</br>
Evidence: His uniform fibers were found at the scene, and he used an emergency override code ("VTK-ADMIN").</h5>

<h4 style="color:red">Reason for Targeting Armstrong</h4>
<h5>Armstrong was killed due to Project Phoenix, a quantum computing initiative with international security implications. Forensic notes reveal he had </br>contingency plans involving offshore accounts and unauthorized technology transfers to foreign entities. The merger with NexusTech threatened to</br> expose vulnerabilities that could compromise national security, making him a target for elimination by both corporate rivals and diplomatic operatives</br>.</h5>

<h4>The murder was a coordinated effort involving Viktor Radek (primary executor), Sarah Chen (financial accomplice), and Marcus Taylor</br>. (security insider), with possible backing from foreign diplomats. The motive was to silence Armstrong and suppress Project Phoenix’s secrets, ensuring </br>control over quantum computing advancements. The use of rare toxins, digital tampering, and diplomatic overrides points to a high-stakes </br>conspiracy blending corporate espionage and international intrigue.</h4>